### PR TITLE
C6: LP_I2C basic driver

### PR DESCRIPTION
**Pull request by JurajSadel**
## Thank you!

Thank you for your contribution.
Please make sure that your submission includes the following:

### Must

- [ ] The code compiles without `errors` or `warnings`.
- [ ] All examples work.
- [ ] `cargo fmt` was run.
- [ ] Your changes were added to the `CHANGELOG.md` in the proper section.
- [ ] You updated existing examples or added examples (if applicable).
- [ ] Added examples are checked in CI

### Nice to have

- [ ] You add a description of your work to this PR.
- [ ] You added proper docs for your newly added features and code.

CC @playfulFence

### CODE REVIEW

File: esp-hal/src/gpio.rs [hunk 1]
```diff
@@ -2735,6 +2739,23 @@ pub mod rtc_io {
                 private: PhantomData,
             }
         }
+
+        #[cfg(esp32c6)]
+        /// Configures the pin as an pullup input and a push pull output pin.
+        pub fn into_puinput_ppoutput(self) -> LowPowerPin<InOut, PIN> {
```
Comment (bjoernQ) [inline esp-hal/src/gpio.rs:2745]: that's a weird name - probably `into_open_drain_output` like we do it for HP GPIO is a better name

File: esp-hal/src/gpio.rs [hunk 2]
```diff
@@ -85,6 +85,8 @@ pub struct PullDown;
 
 pub struct PullUp;
 
+pub struct InOut;
```
Comment (bjoernQ) [inline esp-hal/src/gpio.rs:88]: shouldn't this be `OpenDrain`

File: esp-lp-hal/src/i2c.rs [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+const LP_I2C_NACK_INT_ST_S: u32 = 10;
+
+const I2C_LL_INTR_MASK: u32 = (1 << LP_I2C_TRANS_COMPLETE_INT_ST_S)
+    | (1 << LP_I2C_END_DETECT_INT_ST_S)
+    | (1 << LP_I2C_NACK_INT_ST_S);
+
+const LP_I2C_FIFO_LEN: u32 = 16;
+
+#[doc(hidden)]
+pub unsafe fn conjour() -> Option<LpI2c> {
```
Comment (bjoernQ) [inline esp-lp-hal/src/i2c.rs:16]: a few days ago, I learnt that we should use `conjure` in our codebase

File: esp-lp-hal/src/i2c.rs [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+            .modify(|_, w| w.rx_fifo_rst().set_bit());
+
+        self.i2c
+            .fifo_conf()
+            .modify(|_, w| w.rx_fifo_rst().clear_bit());
+    }
+
+    fn wait_for_completion(&self) -> Result<(), Error> {
+        loop {
+            let interrupts = self.i2c.int_status().read();
```
Comment (bjoernQ) [inline esp-lp-hal/src/i2c.rs:460]: here and probably in other places: usually we use `int_raw` - then we don't need to enable / disable interrupts. is there a reason we cannot use `int_raw` here?

Comment (JurajSadel) [reply esp-lp-hal/src/i2c.rs:460]: I tried to change the interrupts from `status` to `raw` and it reads wrong data 🤷‍♂️

File: esp-lp-hal/src/i2c.rs [hunk 5]
```diff
@@ -13,7 +13,7 @@ const I2C_LL_INTR_MASK: u32 = (1 << LP_I2C_TRANS_COMPLETE_INT_ST_S)
 const LP_I2C_FIFO_LEN: u32 = 16;
 
 #[doc(hidden)]
-pub unsafe fn conjour() -> Option<LpI2c> {
+pub unsafe fn conjure() -> Option<LpI2c> {
```
Comment (MabezDev) [inline esp-lp-hal/src/i2c.rs:16]: Why is this an `Option`?

**Comment by bjoernQ** [comment]
I also tested on HW - apparently it works.

Given the example is not too exciting maybe adding a note that users should use a logic-analyzer or scope to actually see something would be a good thing?

File: esp-lp-hal/src/i2c.rs (hunk 5 above)
Comment (JurajSadel) [reply esp-lp-hal/src/i2c.rs]: TBH I just copied it from GPIO driver - but doesn't make sense to have Option for UART and I2C, removed, thanks!

**Review by bjoernQ — APPROVED** [review]
LGTM - Thanks

**Review by MabezDev — APPROVED** [review]
Am I understanding this correctly that we are copy and pasting the lp driver into esp-hal, once its complete in esp-lp-hal?

Is there any reason we aren't depending on esp-lp-hal in esp-hal and re-exporting the driver there for the HP to use? I guess the pacs have different base addresses.

I will create an issue to fix this at some point, but I'm happy to merge this as is for now.

Thanks @JurajSadel and @playfulFence for working on this and bringing it over the finish line!

**Comment by jessebraham** [comment]
> Am I understanding this correctly that we are copy and pasting the lp driver into esp-hal, once its complete in esp-lp-hal?

Yeah this looks kind of strange, I don't really understand why we're duplicating so much code. This probably needs some refactoring.

**Comment by JurajSadel** [comment]
Merging this so it won't start to rot until we are sure, how to proceed.

**Comment by JurajSadel** [comment]
Thanks for reviews!

**Comment by jessebraham** [comment]
> > Am I understanding this correctly that we are copy and pasting the lp driver into esp-hal, once its complete in esp-lp-hal?
> 
> Yeah this looks kind of strange, I don't really understand why we're duplicating so much code. This probably needs some refactoring.

Or not...

**Pull request merged**